cscs-checks/prgenv/affinity_check.py

```diff
@@ -0,0 +1,155 @@
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+class AffinityTestBase(rfm.RegressionTest):
+    def __init__(self, variant):
+        super().__init__()
+        self.valid_systems = ['daint:gpu', 'daint:mc',
+                              'dom:gpu', 'dom:mc']
+        self.valid_prog_environs = ['PrgEnv-gnu']
+        self.build_system = 'Make'
+        self.build_system.options = ['--directory=affinity', 'MPI=1']
+        self.prebuild_cmd = ['git clone https://github.com/vkarak/affinity']
+        self.executable = 'affinity/affinity'
+        self.variant = variant
+
+        self.maintainers = ['RS', 'VK']
+        self.tags = {'production', 'scs'}
+
+    def setup(self, partition, environ, **job_opts):
+
+        def parse_affinity(x):
+            x_int = [int(xi) for xi in x.split()]
+            return sorted(x_int)
+
+        re_aff_cores = r'CPU affinity: \[\s+(?P<cpu>[\d+\s+]+)\]'
+        self.aff_cores = sn.extractall(
+            re_aff_cores, self.stdout, 'cpu', parse_affinity)
+        self.ref_cores = sn.extractall(
+            re_aff_cores, self.cases[self.variant]['ref_%s' % partition.name],
+            'cpu', parse_affinity)
+        re_aff_thrds = r'^Tag:[^\n\r]*Thread:\s+(?P<thread>\d+)'
+        self.aff_thrds = sn.extractall(re_aff_thrds, self.stdout, 'thread',
+                                       int)
+        self.ref_thrds = sn.extractall(
+            re_aff_thrds, self.cases[self.variant]['ref_%s' % partition.name],
+            'thread', int)
+        re_aff_ranks = r'^Tag:[^\n\r]*Rank:\s+(?P<rank>\d+)[\s+\S+]'
+        self.aff_ranks = sn.extractall(re_aff_ranks, self.stdout, 'rank', int)
+        self.ref_ranks = sn.extractall(
+            re_aff_ranks, self.cases[self.variant]['ref_%s' % partition.name],
+            'rank', int)
+
+        self.use_multithreading = self.cases[self.variant]['multithreading']
+
```

Comment:
It is a bit hard to following what is going here with the sanity checking. What I have understood is that you assume that ranks and threads will be printed in order, so later you can compare the individual sequences, i.e., the sequence of threads, ranks and the corresponding cpu lists. Am I right? Please write a comment here explaining what you are doing, because I believe that there is an assumption here that is not so obvious to follow.